### PR TITLE
[bloomfilter] Refactor bloom filter

### PR DESCRIPTION
**Pull request by Liuhaai**
1. `FromBytes([]byte)` is added into `BloomFilter` interface to replace `BloomFilterFromBytes`. It separates the creation and the data loading of a bloomfilter from `BloomFilterFromBytes`, which enables loads the data from DB directed into existing bloomFilter. https://github.com/iotexproject/go-pkgs/pull/32
  2. Create setter `SetStart()` and `SetEnd()` for struct `bloomRange`. And move the creation of `BloomFilter` into `newBloomRange`.
  3. Replace `rangeBloomFilter` and `getBloomRangeByIndex` with  `loadBloomRangeFromDB`
  4. Delete deprecated interface `RangeFilterByHeight`

### CODE REVIEW

File: blockchain/block/header.go [hunk 1]
```diff
@@ -129,7 +129,15 @@ func (h *Header) loadFromBlockHeaderCoreProto(pb *iotextypes.BlockHeaderCore) er
 	copy(h.deltaStateDigest[:], pb.GetDeltaStateDigest())
 	copy(h.receiptRoot[:], pb.GetReceiptRoot())
 	if pb.GetLogsBloom() != nil {
-		h.logsBloom, err = bloom.BloomFilterFromBytesLegacy(pb.GetLogsBloom(), 2048, 3)
+		// TODO: move the number(2048, 3) to top as const or params of the struct
```
Comment (dustinxie) [inline blockchain/block/header.go:132]: i think adding a comment here documenting that the bloomfilter in header is legacy(2048, 3) would suffice

File: blockindex/bloom_range.go [hunk 2]
```diff
@@ -11,24 +11,30 @@ type bloomRange struct {
 	bloom.BloomFilter
 }
 
-func newBloomRange(start uint64, bf bloom.BloomFilter) *bloomRange {
+func newBloomRange(bfSize, bfNumHash uint64) (*bloomRange, error) {
```
Comment (dustinxie) [inline blockindex/bloom_range.go:14]: good refactor, define it like this, then call `SetStart()` feels better/more clear

File: blockindex/bloomfilterindexer_test.go [hunk 3]
```diff
@@ -79,57 +65,32 @@ func getTestLogBlocks(t *testing.T) []*block.Block {
...
 		SignAndBuild(identityset.PrivateKey(27))
 	require.NoError(t, err)
 
-	hash5 := hash.Hash256{}
-	fnv.New32().Sum(hash5[:])
 	blk5, err := block.NewTestingBuilder().
 		SetHeight(5).
-		SetPrevBlockHash(hash5).
-		SetTimeStamp(testutil.TimestampNow()).
-		AddActions(tsf1, execution1).
```
Comment (dustinxie) [inline blockindex/bloomfilterindexer_test.go:132]: keep these txs in the block

File: go.mod [hunk 4]
```diff
@@ -16,7 +16,7 @@ require (
 	github.com/grpc-ecosystem/go-grpc-prometheus v1.2.0
 	github.com/iotexproject/go-fsm v1.0.0
 	github.com/iotexproject/go-p2p v0.3.3
-	github.com/iotexproject/go-pkgs v0.1.11
+	github.com/iotexproject/go-pkgs v0.1.12-0.20220209063039-b876814568a0
```
Comment (dustinxie) [inline go.mod:19]: v0.1.12 is released

File: blockindex/bloom_range.go [hunk 5]
```diff
@@ -42,23 +48,20 @@ func (br *bloomRange) Bytes() ([]byte, error) {
...
 }
 
-func bloomRangeFromBytes(data []byte) (*bloomRange, error) {
+func (br *bloomRange) FromBytes(data []byte) error {
 	length := len(data)
 	if length <= 16 {
-		return nil, errors.New("not enough data")
+		return errors.New("not enough data")
+	}
+	if br.BloomFilter == nil {
```
Comment (dustinxie) [inline blockindex/bloom_range.go:56]: move this check to beginning?

File: blockindex/bloomfilterindexer.go [hunk 6]
```diff
@@ -190,18 +189,19 @@ func (bfx *bloomfilterIndexer) BlockFilterByHeight(height uint64) (bloom.BloomFi
...
-
-// RangeFilterByHeight returns the range bloomfilter for the height
-func (bfx *bloomfilterIndexer) RangeFilterByHeight(height uint64) (bloom.BloomFilter, error) {
-	br, err := bfx.rangeBloomFilter(height)
+	bf, err := bloom.NewBloomFilter(bfx.bfSize, bfx.bfNumHash)
 	if err != nil {
 		return nil, err
 	}
-	return br.BloomFilter, nil
+	// load data into dummy bloomFilter
```
Comment (dustinxie) [inline blockindex/bloomfilterindexer.go:196]: remove this comment? it is not "dummy" filter?

File: blockindex/bloomfilterindexer.go [hunk 7]
```diff
@@ -297,13 +292,15 @@ func (bfx *bloomfilterIndexer) addLogsToRangeBloomFilter(ctx context.Context, bl
 	}
 }
 
-func (bfx *bloomfilterIndexer) getBloomRangeByIndex(idx uint64) (*bloomRange, error) {
-	bfKey := byteutil.Uint64ToBytesBigEndian(idx)
+func (bfx *bloomfilterIndexer) loadBloomRangeFromDB(br *bloomRange, bfKey []byte) error {
```
Comment (dustinxie) [inline blockindex/bloomfilterindexer.go:295]: i am thinking, is it better to use `uint64 index` as input?

File: blockindex/bloomfilterindexer.go [hunk 8]
```diff
@@ -218,9 +218,13 @@ func (bfx *bloomfilterIndexer) FilterBlocksInRange(l *filter.LogFilter, start, e
...
 	}
+	br, err := newBloomRange(bfx.bfSize, bfx.bfNumHash)
+	if err != nil {
+		return nil, err
+	}
 	for ; startIndex <= endIndex; startIndex++ {
-		br, err := bfx.getBloomRangeByIndex(startIndex)
-		if err != nil {
+		bfKey := byteutil.Uint64ToBytesBigEndian(startIndex)
+		if err := bfx.loadBloomRangeFromDB(br, bfKey); err != nil {
```
Comment (dustinxie) [inline blockindex/bloomfilterindexer.go:227]: so here it's just `loadBloomRangeFromDB(br, startIndex)`

File: blockindex/bloom_range_test.go [hunk 9]
```diff
@@ -3,16 +3,17 @@ package blockindex
...
 	"github.com/stretchr/testify/require"
 )
 
 func TestNewBloomRange(t *testing.T) {
 	require := require.New(t)
 
-	br, err := bloomRangeFromBytes(make([]byte, 16))
-	require.Error(err)
-	require.Nil(br)
+	t.Run("unenough data", func(t *testing.T) {
```
Comment (dustinxie) [inline blockindex/bloom_range_test.go:12]: revert

File: go.mod (hunk 4 above)
Comment (Liuhaai) [reply go.mod:19]:
```
go get github.com/iotexproject/go-pkgs
go get: upgraded github.com/iotexproject/go-pkgs v0.1.12-0.20220209063039-b876814568a0 => v0.1.12
go get: upgraded github.com/iotexproject/iotex-address v0.2.6 => v0.2.7
```
iotex-address will be upgraded.

File: blockindex/bloomfilterindexer.go (hunk 7 above)
Comment (Liuhaai) [reply blockindex/bloomfilterindexer.go:295]: No. See `err := bfx.loadBloomRangeFromDB(bfx.curRangeBloomfilter, bfx.currRangeBfKey);`

Comment (dustinxie) [reply blockindex/bloomfilterindexer.go:293]: i mean pass in index and do the `byteutil.Uint64ToBytesBigEndian` inside

**Review by dustinxie — APPROVED** [review]

File: blockchain/block/header.go [hunk 10]
```diff
@@ -137,7 +137,14 @@ func (h *Header) loadFromBlockHeaderCoreProto(pb *iotextypes.BlockHeaderCore) er
 	copy(h.deltaStateDigest[:], pb.GetDeltaStateDigest())
 	copy(h.receiptRoot[:], pb.GetReceiptRoot())
 	if pb.GetLogsBloom() != nil {
-		h.logsBloom, err = bloom.BloomFilterFromBytesLegacy(pb.GetLogsBloom(), 2048, 3)
+		var err error
```
Comment (CoderZhi) [inline blockchain/block/header.go:140]: delete

File: blockindex/bloomfilterindexer.go [hunk 11]
```diff
@@ -269,6 +260,7 @@ func (bfx *bloomfilterIndexer) commit(blockNumber uint64, blkBloomfilter bloom.B
 	return bfx.kvStore.WriteBatch(b)
 }
 
+// TODO: improve performance
```
Comment (CoderZhi) [inline blockindex/bloomfilterindexer.go:263]: `// TODO (liuhaai): improve performance`

File: blockindex/bloomfilterindexer.go [hunk 12]
```diff
@@ -282,6 +274,7 @@ func (bfx *bloomfilterIndexer) calculateBlockBloomFilter(ctx context.Context, re
 	return bloom
 }
 
+// TODO: improve performance
```
Comment (CoderZhi) [inline blockindex/bloomfilterindexer.go:277]: `// TODO (liuhaai): improve performance`

File: blockindex/bloomfilterindexer.go [hunk 13]
```diff
@@ -190,18 +189,17 @@ func (bfx *bloomfilterIndexer) BlockFilterByHeight(height uint64) (bloom.BloomFi
...
 		return nil, err
 	}
-	return br.BloomFilter, nil
+	if err := bf.FromBytes(bfBytes); err != nil {
+		return nil, err
+	}
+	return bf, nil
 }
 
+// TODO: add pagination in bloomFitler
```
Comment (CoderZhi) [inline blockindex/bloomfilterindexer.go:202]: `// TODO (liuhaai): add pagination in bloomFitler`

**Review by CoderZhi — APPROVED** [review]

File: blockchain/block/header.go (hunk 10 above)
Comment (Liuhaai) [reply blockchain/block/header.go:140]: good finding

File: blockindex/bloomfilterindexer.go (hunk 11 above)
Comment (Liuhaai) [reply blockindex/bloomfilterindexer.go:263]: IMO, this is a good independent task for beginners. I will file it as a good-frist-issue issue.

File: blockindex/bloomfilterindexer.go (hunk 12 above)
Comment (Liuhaai) [reply blockindex/bloomfilterindexer.go:277]: IMO, this is a good independent task for beginners. I will file it as a good-frist-issue issue.

File: blockindex/bloomfilterindexer.go (hunk 13 above)
Comment (Liuhaai) [reply blockindex/bloomfilterindexer.go:202]: ok. Will assign a new issue to myself

**Pull request merged**